app/service/operation_svc.py

```diff
@@ -91,9 +93,30 @@ def __init__(self):
             if agent_output:
                 result = (await self.data_svc.explode_results(criteria=dict(link_id=step['id'])))[0]
                 step_report['output'] = self.decode_bytes(result['output'])
-            report['steps'].append(step_report)
+            agents_steps[step['paw']]['steps'].append(step_report)
+        report['steps'] = agents_steps
+        report['skipped_abilities'] = await self.get_skipped_abilities_by_agent(op_id=op['id'])
         return report
 
+    async def get_skipped_abilities_by_agent(self, op_id):
+        operation = (await self.get_service('data_svc').explode_operation(criteria=dict(id=op_id)))[0]
+        abilities_by_agent = await self._get_all_possible_abilities_by_agent(operation['host_group'],
+                                                                             operation['adversary'])
+        skipped_abilities = []
+        op_results = set([s['ability'] for s in operation['chain']])
+        op_facts = set([f['property'] for f in operation['facts']])
+        # trim to only commands not executed
+        for agent in [a for a in operation['host_group']]:
```

Comment:
should we loop through the abilities_by_agent object, instead of operation['host_group']?

Reply:
I'd challenge you to simplify this function a bit. I think there are a few ways you could tweak it to make it flow more readable

Reply:
Looping over the `abilities_by_agent` object now - but most of the ways to make the flow more readable increase the number of calculations done.  For example:
changing:
```
                skipped = await self._check_reason_skipped(agent=agent,
                                                           ability=ab,
                                                           op_facts=operation_facts,
                                                           op_results=operation_results,
                                                           state=operation['state'],
                                                           agent_executors=agent_executors)
```
to
```
                skipped = await self._check_reason_skipped(agent=agent,
                                                           ability=ab,
                                                           op_facts=set([s['ability'] for s in operation['chain']]),
                                                           op_results=set([f['property'] for f in operation['facts']]),
                                                           state=operation['state'],
                                                           agent_executors=[a['executor'] for a in agent['executors']])
```
Makes the code very clean, but computationally more intensive.

Reply:
> should we loop through the abilities_by_agent object, instead of operation['host_group']?

I need to loop over the host_group because that contains the entire exploded agent information whereas the abilities_by_agent is just a quick dictionary to avoid having to loop over the host_group to do inserts.